zend_abstract_interface/config/config_decode.c

```diff
@@ -0,0 +1,201 @@
+#include "config_decode.h"
+
+#include <assert.h>
+#include <ctype.h>
+#include <main/php.h>
+#include <stdbool.h>
+#include <string.h>
+#include <strings.h>
+
+#include "config.h"
+
+static bool zai_config_decode_bool(zai_string_view value, zval *decoded_value) {
+    if (value.len == 1) {
+        if (strcmp(value.ptr, "1") == 0) {
+            ZVAL_TRUE(decoded_value);
+        } else if (strcmp(value.ptr, "0") == 0) {
+            ZVAL_FALSE(decoded_value);
+        }
+    } else if ((value.len == 4 && strcasecmp(value.ptr, "true") == 0)) {
+        ZVAL_TRUE(decoded_value);
+    } else if ((value.len == 5 && strcasecmp(value.ptr, "false") == 0)) {
+        ZVAL_FALSE(decoded_value);
+    }
+    return Z_TYPE_P(decoded_value) == IS_TRUE || Z_TYPE_P(decoded_value) == IS_FALSE;
+}
+
+/* Since strtod() supports conversion of constants like 'NaN' and 'INF', as well
+ * as successfully decoding values such as '4foo', we check that only numbers,
+ * decimal points, and whitespace exist before passing to strtod().
+ */
+static bool zai_config_is_valid_double_format(const char *str) {
+    bool seen_decimal = false;
+    while (*str) {
+        if (!isdigit(*str) && !isspace(*str) && *str != '.') return false;
+        if (*str == '.') {
+            /* strtod will convert '4.2.0' to a double, so we block it. */
+            if (seen_decimal) return false;
+            seen_decimal = true;
+        }
+        str++;
+    }
+    return true;
+}
+
+static bool zai_config_decode_double(zai_string_view value, zval *decoded_value) {
+    if (!zai_config_is_valid_double_format(value.ptr)) return false;
+
+    char *endptr = (char *)value.ptr;
+
+    // The strtod function is a bit tricky, so I've quoted docs to explain code
+
+    /* Since 0 can legitimately be returned on both success and failure, the
+     * calling program should set errno to 0 before the call, and then
+     * determine if an error occurred by checking whether errno has a nonzero
+     * value after the call.
+     */
+    errno = 0;
+    double result = strtod(value.ptr, &endptr);
+
+    /* If endptr is not NULL, a pointer to the character after the last
+     * character used in the conversion is stored in the location referenced
+     * by endptr. If no conversion is performed, zero is returned and the value
+     * of nptr is stored in the location referenced by endptr.
+     */
+    if (endptr != value.ptr && errno == 0) {
+        ZVAL_DOUBLE(decoded_value, result);
+        return true;
+    }
+    return false;
+}
+
+/* As with strtod(), strtoll() also supports conversion of constants and
+ * ill-formatted numbers so we add a number-format check.
+ */
+static bool zai_config_is_valid_int_format(const char *str) {
+    while (*str) {
+        if (!isdigit(*str) && !isspace(*str)) return false;
+        str++;
+    }
+    return true;
+}
+
+static bool zai_config_decode_int(zai_string_view value, zval *decoded_value) {
```

Comment:
maybe just use `zend_atol` to mirror `ZEND_INI_MH(OnUpdateLong)` behaviour?

Reply:
... We seem to be validating values here, which is quite atypical for inis in PHP.  In PHP inis are generally only supposed to fail if the value cannot be worked with. (like ini value expects number > 0 and you get -1.)
I'm not sure whether we want to break with that.

Reply:
This all sounds good to me. If we end up tweaking the behavior of env var parsing to be more in line with INI, I think we should go for it. That is, unless there is an obvious BC or something. :)